### PR TITLE
Check a subscription's endpoint unicity before adding it

### PR DESCRIPTION
**Pull request by julienw**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -45,9 +45,17 @@ app.use((err, req, res, _next) => {
...
+    error: err.name,
+    code: err.code,
+    message: err.message,
+  };
+
+  if (err.data) {
+    errorMessage.data = err.data;
+  }
+
+  res.status(err.status || 500).send(errorMessage);
```
Comment (julienw) [inline app.js:58]: This allows to add some generic data to an error.

File: dao/database.js [hunk 2]
```diff
@@ -134,7 +134,7 @@ const createStatements = [`
     id INTEGER PRIMARY KEY AUTOINCREMENT,
     family TEXT,
     title TEXT,
-    endpoint TEXT,
+    endpoint TEXT UNIQUE,
```
Comment (julienw) [inline dao/database.js:137]: Ensures database consistency.

File: routes/subscriptions.js [hunk 3]
```diff
@@ -13,15 +14,33 @@ function hidePrivateData(item) {
...
+    return subscriptions.create(family, req.body).then((id) => {
+      debug('Subscription #%s has been created in database', id);
+
+      return subscriptions.show(family, id);
+    }).then((subscription) => {
+      res
+        .status(201)
+        .location(`${req.baseUrl}/${subscription.id}`)
+        .send(hidePrivateData(subscription));
+    });
```
Comment (julienw) [inline routes/subscriptions.js:43]: here we start having some complexity that would justify a separate "service" layer :)

File: routes/subscriptions.js [hunk 4]
```diff
@@ -13,15 +14,33 @@ function hidePrivateData(item) {
...
-  }).then((subscription) => {
-    res
-      .status(201)
-      .location(`${req.baseUrl}/${subscription.id}`)
-      .send(hidePrivateData(subscription));
+
+  const endpoint = req.body.subscription.endpoint;
+  subscriptions.findByEndpoint(family, endpoint).then((subscription) => {
+    // This endpoint is already in the DB
+    throw new DuplicateEndpointError( // will generate a status "409"
```
Comment (julienw) [inline routes/subscriptions.js]: "409" is "Conflict"

File: utils/errors.js [hunk 5]
```diff
@@ -1,9 +1,12 @@
 function createError(status, name) {
-  function NewError(code, message) {
+  function NewError(code, message, data) {
     this.code = code;
     this.message = message;
     this.status = status;
     this.name = name;
+    if (data) {
+      this.data = data;
+    }
```
Comment (julienw) [inline utils/errors.js:9]: I didn't want to use simply `this.data = data` (without the condition) just by correctness.

File: dao/subscriptions.js [hunk 6]
```diff
@@ -11,6 +11,13 @@ function notFoundError(id) {
   );
 }
 
+function endpointNotFoundError(endpoint) {
```
Comment (JohanLorenzo) [inline dao/subscriptions.js:14]: On a middle-term perspective, I wonder if we shouldn't factorize this type of function. Something like:

``` js
function notFoundError(subject, discriminant) {
  return new NotFoundError(
    `${subject}_not_found`,
    `The ${subject} with ${discriminant.name} \`${discriminant.value}\` does not exist.`
}
```

What do you think?

File: routes/subscriptions.js (hunk 4 above)
Comment (JohanLorenzo) [inline routes/subscriptions.js]: Thanks for the reminder 😃 . Nit: I'd add "Conflict" in the code comment, for next readers.

File: routes/subscriptions.js (hunk 3 above)
Comment (JohanLorenzo) [inline routes/subscriptions.js:43]: Yeah, I'd prefer not keeping the current complexity. I'm fine landing it as is for now, but a service layer like `subscriptionsService.verifyEndpointDoesNotExist()` would be great. We could have this route as small as:

``` js
const family = req.user.family;
const endpoint = req.body.subscription.endpoint;

return subscriptionsService.verifyEndpointDoesNotExist(family, endpoint)
  .then(() => subscriptionsService.create(family, req.body))
  .then((subscription) => {
    res
      .status(201)
      .location(`${req.baseUrl}/${subscription.id}`)
      .send(subscription) // hide private data is done in the service
});
```

**Comment by JohanLorenzo** [comment]
r+. Let's make the service layer in another PR. One tiny nit to change, so we don't forget about it.

File: dao/subscriptions.js (hunk 6 above)
Comment (julienw) [inline dao/subscriptions.js:14]: filed https://github.com/fxbox/calendar-server/issues/26

File: routes/subscriptions.js (hunk 3 above)
Comment (julienw) [inline routes/subscriptions.js:43]: filed https://github.com/fxbox/calendar-server/issues/27

**Pull request merged**